libraries/common/io.effekt

```diff
@@ -6,49 +6,80 @@ import queue
 extern include llvm "../llvm/io.ll"
 extern include js "../js/io.js"
 
-// Asynchronous I/O
-// ----------------
-
-interface AsyncIO {
-  /**
-   * Runs the provided computation and conceptually waits until it completes by
-   * calling the writer-continuation.
-   *
-   * The provided writer-continuation should only ever be called ONCE.
-   * The computation is passed as a thunk in order to prevent accidental usage of
-   * other control effects that would violate the contract of calling the writer
-   * once.
-   */
-  def performIO[T](computation: (T => Unit at {io, global}) => Unit at {io, global}): T
-}
 
-/**
-  * Run an IO computation (that uses callbacks) here.
-  */
-def blocking(program: () => Unit / AsyncIO at {io, global}): Unit =
-  try { program() }
-  with AsyncIO {
-    def performIO[T](computation) = computation(resume)
-  }
+// Event Loop
+// ----------
+
+interface Eventloop {}
+
+type Task[T] = () => T / Eventloop at {io, control, global}
+
+def eventloop(program: Task[Unit]): Unit = direct::eventloop(box {
+  try program() with Eventloop {}
+})
+
+def spawn(task: Task[Unit]): Unit / Eventloop =
+  // TODO installing a handler here is unnecessary
+  direct::spawn(box{ try task() with Eventloop {} })
+
+def yield(): Unit / Eventloop =
+  direct::yield()
+
+def abort(): Nothing / Eventloop =
+  direct::abort()
+
+namespace direct {
+
+  extern io def eventloop(program: () => Unit at {global,io,control}): Unit =
+  js "${program}().run()"
+  llvm """
+    call void @run(%Neg %program)
+    %loop = call ptr @uv_default_loop()
+    call i32 @uv_run(ptr %loop, i32 0)
+    call void @uv_loop_close(ptr %loop)
+    ret %Pos zeroinitializer
+  """
+
+  extern control def spawn(task: () => Unit at {global,io,control}): Unit =
+    js "$effekt.callcc(callback => { setTimeout(() => callback($effekt.unit), 0); ${task}().run()})"
+    llvm """
+      call void @c_timer_start(%Int 0, %Stack %stack)
+      call void @run(%Neg %task)
+      ret void
+    """
 
+  extern control def yield(): Unit =
+    js "$effekt.callcc(callback => setTimeout(() => callback($effekt.unit), 0))"
+    llvm """
+      call void @c_timer_start(%Int 0, %Stack %stack)
+      ret void
+    """
+
+  extern control def abort(): Nothing =
+    js "$effekt.callcc(callback => { return })"
+    llvm """
+      call void @eraseStack(%Stack %stack)
+      ret void
+    """
 
-interface Concurrent {
-  def yield(): Unit
-  def abort(): Nothing
 }
 
 // Promises
 // --------
 
 extern type Promise[T]
-  // = js "{promise: Promise, resolve: ƒ}"
-  // = llvm "Promise"
+  // = js "{resolve: ƒ, promise: Promise}"
+  // = llvm "{tag: 0, obj: Promise*}"
 
-interface Promises {
-  def promise[T](p: Task[T]): Promise[T]
-  def await[T](p: Promise[T]): T
+def promise[T](task: Task[T]): Promise[T] / Eventloop = {
+  val p = promise::make[T]();
+  spawn(box { promise::resolve(p, task()) });
+  return p
 }
```

Comment:
I think it is fine, but please be aware that with all of these changes the scheduling and semantics of promises is baked into the language and cannot be changed by users.